_posts/2022-05-09-project-loom-for-distributed-systems.md

```diff
@@ -0,0 +1,397 @@
+assertThat(executor).hasSize(1)
+    .as("runnable for vthread has been submitted");
+executor.poll().run();
+assertThat(executor).hasSize(0)
+    .as("vthread has blocked, no longer runnable");
+lock.unlock();
+assertThat(executor).hasSize(1)
+    .as("due to unlock, the vthread is now schedulable again");
+executor.poll().run();
+assertThat(lock.tryLock()).isFalse()
+    .as("the virtual thread now holds the lock");
+```
+
+By utilizing this API, we can exert fine grained deterministic control over execution within Java. Let's give a micro level example. Suppose we're trying to test the correctness of a buggy version of Guava's `Suppliers.memoize` function.
+
+```java
+class BrokenMemoizingSupplier<T> implements Supplier<T> {
+  private final Lock lock;
+  private final Supplier<T> delegate;
+  private volatile boolean initialized = false;
+  private volatile T value;
+
+  BrokenMemoizingSupplier(Supplier<T> delegate) {
+    this(new ReentrantLock(), delegate);
+  }
+
+  @VisibleForTesting
+  BrokenMemoizingSupplier(Lock lock, Supplier<T> delegate) {
+    this.lock = lock;
+    this.delegate = delegate;
+  }
+
+  @Override
+  public T get() {
+    if (!initialized) {
+      lock.lock();
+      try {
+        // This code is broken because initialized may have changed between its check before the call to 'lock' and now.
+        T result = delegate.get();
+        value = result;
+        initialized = true;
+        return result;
+      } finally {
+        lock.unlock();
+      }
+    }
+    return value;
+  }
+}
+```
+
+This test is enough to deterministically exhibit the bug:
+
+```java
+// This will fail every time
+@ParameterizedTest
+@MethodSource("range")
+void testAllCases(int seed) {
+  AtomicInteger countCalls = new AtomicInteger();
+  BrokenMemoizingSupplier<Integer> supplier =
+    new BrokenMemoizingSupplier<>(
+      new YieldingLock(),
+      countCalls::incrementAndGet);
+  Random random = new Random(seed);
+  RandomizingExecutor executor = new RandomizingExecutor(random);
+  newVirtualThread(executor, supplier::get);
+  newVirtualThread(executor, supplier::get);
+  executor.drain();
+  assertThat(supplier.get()).isOne();
+}
+
+static Stream<Integer> range() {
+  return IntStream.range(0, 10).boxed();
+}
+
+/**
+  * This executor executes randomly for brevity. By tracking which
+  * tasks launch which other tasks, one could generate all valid
+  * interleavings for small cases such as this.
+  */
+private static final class RandomizingExecutor implements Executor {
+  private final Random random;
+  private final List<Runnable> queue = new ArrayList<>();
+
+  private RandomizingExecutor(Random random) {
+    this.random = random;
+  }
+
+  @Override
+  public void execute(Runnable command) {
+    queue.add(command);
+  }
+
+  public void drain() {
+    while (!queue.isEmpty()) {
+      Collections.shuffle(queue, random);
+      Runnable task = queue.remove(queue.size() - 1);
+      task.run();
```

Comment:
am I understanding correctly that the runnable's `Thread.yield()` will put a new Runnable on the same executor (ie, `this` in this context) and thus add a new runnable onto the `this.queue` whose next instruction is `super.lock()`?